### DISCUSSION TITLE
SQL Query mengambil data terupdate

### DISCUSSION THREAD

**Discussion by Antok1989** [teamalgoritma/community#510] [Q&A] [answered]
Dear Team Algoritma,

Saya ada problem mengenai penarikan data menggunakan SQL. Misal saya mempunyai 2 table, table pertama berisikan data client. Sedangkan table kedua berisikan data transaksi.

![image](https://user-images.githubusercontent.com/120188071/213198614-18c5a810-8610-4690-8ff5-932bcbb323ee.png)

Jika menggunakan fitur **GROUP BY**, maka value untuk masing-masing **charge_code** akan terjumlah.
Jika menggunakan fungsi **SORT** di kolom **Last_Modified**, tetap muncul 6 line hanya saja urutannya terbalik.

Terima kasih

**Comment by ramdhinafinita** [comment]
Dear Pak @Antok1989 

Terima kasih atas pertanyaan yang disampaikan, pertanyaan Anda sudah kami terima. Kami akan mencoba membantu mencarikan referensi berkaitan dengan pertanyaan yang diberikan.

**Comment by LyciaPricia** [comment]
mungkin pakai LASTMODIFIED mas Febrianto

SELECT * FROM dbo.SampleTable 
WHERE [LastModified] = (SELECT MAX([LastModified]) FROM dbo.SampleTable)

sehingga akan mengambil data yang dimodified terakhir.

**Reply by IrfanChairurrachman** [reply]
Dear Pak @Antok1989 

Di sini kami mencoba membantu berdasarkan jawaban dari @LyciaPricia karena secara logic sudah sangat betul krn kita bisa menggunakan subquery untuk mendapatkan tanggal paling maksimal untuk tabel sumber. Namun, bisa saja tiap vouchernya memiliki date maksimal yang berbeda-beda. Untuk mendapatkan tanggal maksimal dari setiap vouchernya bisa dengan `SELECT MAX(LastModified) FROM dbo.SampleTable GROUP BY Code_voucher`.
Sehingga jika kita improve dari query di atas akan menjadi
```
SELECT * FROM dbo.SampleTable
WHERE [LastModified] IN (SELECT MAX(LastModified) FROM dbo.SampleTable GROUP BY Code_voucher)
```

query ini masih dapat diimprove sehingga silahkan jika ada yang mempunyai solusi lebih baik kita dapat diskusi bersama.

Sekian
Team Algoritma

**Comment by ELITSZORD** [comment, marked as answer]
Dear Pak @Antok1989 ,

Izin coba membantu terkait pertanyaan bapak. Kalau ada kesalahan atas jawabannya, please let me know ya pak @Antok1989 .

**# Cara 1**

```
SELECT code_voucher, t1.client, t2.charge_code, t2.value 
FROM Table1 as t1
LEFT JOIN Table2 as t2
WHERE  last_modified IN (SELECT  MAX(last_modified)
				    FROM Table2 
				    GROUP BY charge_code, last_modified  
				    ORDER BY last_modified DESC)
LIMIT 3; 
```
- Jika menggunakan querry diatas, saya ragu di querry MAX yang kemungkinan hanya memunculkan 1 value saja, sehingga jika untuk setiap charge_code memiliki last_modified di hari yang berbeda tidak muncul. 

**# Cara 2**

```
WITH df_1  AS (SELECT *
		      FROM Table1 as t1
		      LEFT JOIN Table2 as t2 
		      USING (code_voucher)
		      GROUP BY charge_code, last_modified  
		      ORDER BY last_modified DESC)
SELECT df_1.code_voucher, df_1.client, df_1.charge_code, df_1.value
FROM df_1
LIMIT 3;
```

**# Cara 3**

```
SELECT code_voucher, t1.client, t2.charge_code, t2.value
FROM Table1 as t1
LEFT JOIN Table2 as t2 
ON t1.code_voucher = t2.code_voucher 
GROUP BY charge_code, last_modified  
ORDER BY last_modified DESC
LIMIT 3;
```
- Untuk date (last_modified) mungkin bisa lebih di detailkan (ada waktunya), sehingga ketika di suatu tanggal tertentu ada beberapa code (charge_code) bisa di groupping hingga ke detail date yang second / milisecondnya.

- Limit 3 disesuaikan dengan jumlah jenis code (charge_code) untuk ambil yang teratas/terakhir last_modifiednya.  

Best Regards.

**Reply by IrfanChairurrachman** [reply]
Dear Pak @ELITSZORD ,

Terima kasih atas masukan-masukan solusinya. Betul bahwa masih terdapat kekurangan dari solusi yang ajukan berhubung hanya mengembalikan 1 nilai tanggal, dimana bisa saja last_modified untuk masing-masing voucher bisa saja memiliki tanggal yang berbeda-beda. 
Untuk solusi yang Bapak ajukan (Solusi 2 dan 3 yang kurang lebih sama) juga dapat menjadi opsi juga dengan catatan LIMITnya harus disesuaikan dengan jumlah row sehingga setiap vouchernya dapat terambil.

Sekian
